File: llm/tests/test_agent.py
"""Tests for llm/agent.py"""

import pytest
import llm.agent as agent
import re


class TestConstants:
    def test_workspace(self):
        assert agent.WORKSPACE.endswith(".agent0")  # Expanded path

    def test_model(self):
        assert agent.MODEL == "minimax-m2.5:cloud"

    def test_max_turns(self):
        assert agent.MAX_TURNS == 5


class TestSystemPrompt:
    def test_system_prompt_exists(self):
        assert len(agent.SYSTEM_PROMPT) > 0
        assert "<shell>" in agent.SYSTEM_PROMPT
        assert "<end/>" in agent.SYSTEM_PROMPT


class TestBuildContext:
    def test_build_context_empty(self):
        """Test build_context with no history"""
        # Save original state
        orig_history = agent.conversation_history.copy()
        orig_key = agent.key_info.copy()
        
        try:
            agent.conversation_history = []
            agent.key_info = []
            result = agent.build_context()
            assert result == ""
        finally:
            # Restore
            agent.conversation_history = orig_history
            agent.key_info = orig_key

    def test_build_context_with_memory(self):
        """Test build_context with key_info"""
        orig_history = agent.conversation_history.copy()
        orig_key = agent.key_info.copy()
        
        try:
            agent.conversation_history = []
            agent.key_info = ["重要資訊1", "重要資訊2"]
            result = agent.build_context()
            assert "<memory>" in result
            assert "重要資訊1" in result
        finally:
            agent.conversation_history = orig_history
            agent.key_info = orig_key

    def test_build_context_with_history(self):
        """Test build_context with conversation history"""
        orig_history = agent.conversation_history.copy()
        orig_key = agent.key_info.copy()
        
        try:
            agent.conversation_history = [
                "<user>Hello</user>",
                "<assistant>Hi</assistant>",
            ]
            agent.key_info = []
            result = agent.build_context()
            assert "<history>" in result
        finally:
            agent.conversation_history = orig_history
            agent.key_info = orig_key


class TestUpdateMemory:
    def test_update_memory_basic(self):
        """Test update_memory adds to history"""
        orig = agent.conversation_history.copy()
        
        try:
            agent.conversation_history = []
            agent.update_memory("Hello", "Hi there!", None)
            assert len(agent.conversation_history) == 2
        finally:
            agent.conversation_history = orig

    def test_update_memory_with_tool_result(self):
        """Test update_memory with tool result"""
        orig = agent.conversation_history.copy()
        
        try:
            agent.conversation_history = []
            agent.update_memory("Run cmd", "Output", "tool result")
            assert len(agent.conversation_history) == 3
            assert "<tool>" in agent.conversation_history[-1]
        finally:
            agent.conversation_history = orig

    def test_update_memory_truncates(self):
        """Test update_memory truncates long history"""
        orig = agent.conversation_history.copy()
        
        try:
            # Add 20 items (MAX_TURNS * 4 = 20)
            agent.conversation_history = []
            for i in range(20):
                agent.conversation_history.append(f"<user>msg{i}</user>")
            
            agent.update_memory("new", "response", None)
            # Should not exceed MAX_TURNS * 4 = 20
            assert len(agent.conversation_history) <= 20
        finally:
            agent.conversation_history = orig


class TestExtractKeyInfo:
    def test_extract_key_info_regex(self):
        """Test the regex pattern for extracting key info"""
        result = """<memory>
  <item>記住用戶喜歡貓</item>
  <item>用戶住在台北</item>
</memory>"""
        matches = re.findall(r'<item>(.*?)</item>', result, re.DOTALL)
        assert len(matches) == 2

    def test_extract_key_info_empty(self):
        """Test empty memory output"""
        result = "<memory></memory>"
        matches = re.findall(r'<item>(.*?)</item>', result, re.DOTALL)
        assert len(matches) == 0


class TestShellTagParsing:
    def test_parse_single_shell(self):
        """Test parsing single shell command"""
        response = "Hello <shell>echo test</shell> world"
        matches = re.findall(r'<shell>(.+?)</shell>', response, re.DOTALL)
        assert len(matches) == 1
        assert matches[0].strip() == "echo test"

    def test_parse_multiple_shell(self):
        """Test parsing multiple shell commands"""
        response = "<shell>ls</shell> then <shell>cat file.txt</shell>"
        matches = re.findall(r'<shell>(.+?)</shell>', response, re.DOTALL)
        assert len(matches) == 2

    def test_parse_no_shell(self):
        """Test with no shell tags"""
        response = "Just text"
        matches = re.findall(r'<shell>(.+?)</shell>', response, re.DOTALL)
        assert len(matches) == 0


class TestEndTag:
    def test_has_end_tag_true(self):
        """Test end tag detection"""
        response = "Some text <end/> more"
        assert "<end/>" in response

    def test_has_end_tag_false(self):
        """Test no end tag"""
        response = "No end tag here"
        assert "<end/>" not in response


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: llm/tests/test_agent.rs
//! llm/tests/test_agent.rs - AI Agent tests.

use crate::llm::agent::Agent;

#[test]
fn test_agent_creation() {
    let agent = Agent::new();
    let _context = agent.build_context();
    assert!(true);
}

#[test]
fn test_build_context() {
    let agent = Agent::new();
    let context = agent.build_context();
    assert!(context.is_empty());
}

#[test]
fn test_build_context_with_memory() {
    let mut agent = Agent::new();
    agent.add_key_info("重要資訊1".to_string());
    agent.add_key_info("重要資訊2".to_string());

    let context = agent.build_context();
    assert!(context.contains("<memory>"));
    assert!(context.contains("重要資訊1"));
}

#[test]
fn test_build_context_with_history() {
    let mut agent = Agent::new();
    agent.update_memory("Hello", "Hi there!", None);

    let context = agent.build_context();
    assert!(context.contains("<history>"));
}

#[test]
fn test_update_memory() {
    let mut agent = Agent::new();
    agent.update_memory("Hello", "Hi there!", None);
    assert_eq!(agent.show_memory().len(), 0);
}

#[test]
fn test_add_key_info() {
    let mut agent = Agent::new();
    agent.add_key_info("Important info".to_string());
    assert_eq!(agent.show_memory().len(), 1);
}

#[test]
fn test_parse_shell_tags() {
    let response = "Hello <shell>echo test</shell> world";
    let cmds = Agent::parse_shell_tags(response);
    assert_eq!(cmds.len(), 1);
    assert_eq!(cmds[0], "echo test");
}

#[test]
fn test_parse_multiple_shell() {
    let response = "<shell>ls</shell> then <shell>cat file.txt</shell>";
    let cmds = Agent::parse_shell_tags(response);
    assert_eq!(cmds.len(), 2);
}

#[test]
fn test_parse_no_shell() {
    let response = "Just text";
    let cmds = Agent::parse_shell_tags(response);
    assert_eq!(cmds.len(), 0);
}

#[test]
fn test_has_end_tag() {
    assert!(Agent::has_end_tag("Hello <end/>"));
    assert!(!Agent::has_end_tag("Hello"));
}

#[test]
fn test_execute_shell() {
    let agent = Agent::new();
    let result = agent.execute_shell("echo hello");
    assert!(result.is_ok());
    assert!(result.unwrap().contains("hello"));
}

#[test]
fn test_constants() {
    use crate::llm::agent;
    assert_eq!(agent::WORKSPACE, "~/.agent0");
    assert_eq!(agent::MODEL, "minimax-m2.5:cloud");
    assert_eq!(agent::MAX_TURNS, 5);
}